vllm/model_executor/models/eagle.py

```diff
@@ -0,0 +1,86 @@
+from typing import Iterable, List, Optional, Tuple
+
+import torch
+import torch.nn as nn
+
+from vllm.attention.backends.abstract import AttentionMetadata
+from vllm.model_executor.model_loader.weight_utils import default_weight_loader
+from vllm.model_executor.models import ModelRegistry
+from vllm.model_executor.sampling_metadata import SamplingMetadata
+from vllm.sequence import IntermediateTensors, SamplerOutput
+from vllm.transformers_utils.configs.eagle import EAGLEConfig
+
+
+class EAGLE(nn.Module):
+
+    def __init__(self, config: EAGLEConfig, *args, **kwargs) -> None:
+        super().__init__()
+        self.config = config
+
+        architectures = getattr(self.config.model, "architectures", [])
+        for arch in architectures:
+            model_cls = ModelRegistry.load_model_cls(arch)
+            if model_cls is not None:
+                break
+
+        self.model = model_cls(self.config.model, *args, **kwargs)
+        self.fc = nn.Linear(config.model.hidden_size * 2,
+                            config.model.hidden_size,
+                            bias=False)
+
+        self.token_map = None
+
+    @property
+    def sampler(self):
+        return self.model.sampler
+
+    def forward(
```

Comment:
Wondering if you have been able to run any comparison between eagle and medusa and how the results compare?

Reply:
Reopening this comment. Wondering if you have been able to compare the performance of the eagle model with that of medusa?

Reply:
Sorry, I thought I already replied to this.

Yes, I compared the two for Meta-llama-3-8b but found that the overhead for speculation is currently very high for EAGLE. Because of this, the performance of EAGLE (from https://huggingface.co/yuhuili/EAGLE-LLaMA3-Instruct-8B) was similar to that of without any speculation even though the average tokens/step was around 1.25 on MT-Bench.

I am also trying to run a more comprehensive benchmarking or no spec vs Medusa vs EAGLE and will share the results when complete.